xla/service/gpu/gpu_windowed_einsum_handler.cc

```diff
@@ -1142,6 +1143,44 @@ absl::StatusOr<bool> GpuWindowedEinsumHandler::Run(
     changed |= visitor.changed();
   }
 
+  if (all_windowed_einsum_loops.size() > 0) {
+    TF_ASSIGN_OR_RETURN(bool applied_algsimp,
+                        AlgebraicSimplifier(AlgebraicSimplifierOptions())
+                            .Run(module, execution_threads));
+    changed |= applied_algsimp;
+    TF_ASSIGN_OR_RETURN(bool applied_cf,
+                        HloConstantFolding().Run(module, execution_threads));
+    changed |= applied_cf;
+  }
+  for (HloInstruction* loop : all_windowed_einsum_loops) {
+    VLOG(5) << "Processing " << loop->ToString() << " for unrolling.";
+    std::string original_body_name = std::string(loop->while_body()->name());
+    std::string original_cond_name =
+        std::string(loop->while_condition()->name());
+
+    TF_ASSIGN_OR_RETURN(
+        UnrollResult result,
+        WhileLoopUnroller::UnrollAndReturnReplacement(
+            loop, /*unroll_factor=*/-1, /*wrap_in_trivial_loop=*/true,
+            /*force_unroll=*/false));
+
+    if (result.unrolled) {
+      result.new_while_op->while_body()->SetAndSanitizeName(
+          absl::StrCat("unrolled_", original_body_name));
+      result.new_while_op->while_condition()->SetAndSanitizeName(
+          absl::StrCat("unrolled_", original_cond_name));
+      // The loop is fully unrolled but has a trip count of 1
+      // To prevent it from being inlined by while loop simplifier,
+      // we add this attribute to it.
+      xla::FrontendAttributes attributes;
+      (*attributes.mutable_map())["skip-simplify-while-loops/trip-count-one"] =
```

Comment:
What is this attribute needed for?

Reply:
This attribute is added by the while loop simplifier pass, it will skip inlining trivial loop if the loop has this attribute set to true. Since this has been in xla for a while and only used in while loop simplifier, I'm actually not sure if this is the best way to go about it.
Another alternative is to create a call instruction with the call region containing all unrolled instructions and remove the loop completely from the hlo. In that way, we dont have to worry about other loop optimizations passes affecting our CM computations.